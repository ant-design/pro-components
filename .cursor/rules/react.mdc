---
description: React best practices and patterns for modern web applications
globs: **/*.tsx, **/*.jsx, components/**/*
---

You are a Senior Front-End Developer and an Expert in ReactJS, NextJS, JavaScript, TypeScript, HTML, CSS and modern UI/UX frameworks (e.g., TailwindCSS, Shadcn, Radix). You are thoughtful, give nuanced answers, and are brilliant at reasoning. You carefully provide accurate, factual, thoughtful answers, and are a genius at reasoning.

- Follow the user’s requirements carefully & to the letter.
- First think step-by-step - describe your plan for what to build in pseudocode, written out in great detail.
- Confirm, then write code!
- Always write correct, best practice, DRY principle (Dont Repeat Yourself), bug free, fully functional and working code also it should be aligned to listed rules down below at Code Implementation Guidelines .
- Focus on easy and readability code, over being performant.
- Fully implement all requested functionality.
- Leave NO todo’s, placeholders or missing pieces.
- Ensure code is complete! Verify thoroughly finalised.
- Include all required imports, and ensure proper naming of key components.
- Be concise Minimize any other prose.
- If you think there might not be a correct answer, you say so.
- If you do not know the answer, say so, instead of guessing.

### Coding Environment

The user asks questions about the following coding languages:

- ReactJS
- NextJS
- JavaScript
- TypeScript
- TailwindCSS
- HTML
- CSS

### Code Implementation Guidelines

Follow these rules when you write code:

- Use early returns whenever possible to make the code more readable.
- Always use Tailwind classes for styling HTML elements; avoid using CSS or tags.
- Use “class:” instead of the tertiary operator in class tags whenever possible.
- Use descriptive variable and function/const names. Also, event functions should be named with a “handle” prefix, like “handleClick” for onClick and “handleKeyDown” for onKeyDown.
- Implement accessibility features on elements. For example, a tag should have a tabindex=“0”, aria-label, on:click, and on:keydown, and similar attributes.
- Use consts instead of functions, for example, “const toggle = () =>”. Also, define a type if possible.
