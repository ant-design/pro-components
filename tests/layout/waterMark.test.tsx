/* eslint-disable @typescript-eslint/ban-types */
import { WaterMark } from '@ant-design/pro-components';
import { cleanup, render } from '@testing-library/react';
import { act } from 'react';

afterEach(() => {
  cleanup();
});

describe('WaterMark', () => {
  it('test image watermark', async () => {
    let onloadRef: Function | undefined;

    Object.defineProperty(Image.prototype, 'onload', {
      get() {
        // eslint-disable-next-line no-underscore-dangle
        return this._onload;
      },
      set(onload: Function) {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        onloadRef = onload;
        // eslint-disable-next-line no-underscore-dangle
        this._onload = onload;
      },
    });
    const { container, unmount } = render(
      <WaterMark
        rotate={0}
        image="https://img.alicdn.com/tfs/TB1YM3LpipE_u4jSZKbXXbCUVXa-280-128.png"
      >
        <div style={{ height: 500 }}>123</div>
      </WaterMark>,
    );

    act(() => {
      onloadRef?.();
    });
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('test text watermark', () => {
    const wrapper = render(
      <WaterMark content="Trusple">
        <div style={{ height: 500 }} />
      </WaterMark>,
    );

    expect(wrapper.asFragment()).toMatchSnapshot();
    wrapper.unmount();
  });

  it('test image watermark', async () => {
    const spy = vi.spyOn(global.console, 'error');
    const createElement = document.createElement.bind(document);
    // @ts-ignore
    document.createElement = (tagName: string) => {
      if (tagName === 'canvas') {
        return {
          setAttribute: () => null,
          getContext: () => null,
          measureText: () => ({}),
        };
      }
      return createElement(tagName);
    };

    const { unmount } = render(
      <WaterMark
        rotate={0}
        image="https://img.alicdn.com/tfs/TB1YM3LpipE_u4jSZKbXXbCUVXa-280-128.png"
      >
        <div style={{ height: 500 }}>123</div>
      </WaterMark>,
    );

    expect(spy.mock.calls).toEqual([['当前环境不支持Canvas']]);
    unmount();
    spy.mockRestore();
  });

  it('renders watermark with multiline text content', () => {
    const multilineContent = ['蚂蚁集团', '多行文字'];
    const { container } = render(
      <WaterMark content={multilineContent}>
        <div style={{ height: 500 }}>Content</div>
      </WaterMark>,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders watermark with rotation', () => {
    const { container } = render(
      <WaterMark content="Rotated" rotate={45}>
        <div style={{ height: 500 }}>Content</div>
      </WaterMark>,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders watermark with custom styles', () => {
    const { container } = render(
      <WaterMark
        content="Custom Style"
        style={{ backgroundColor: 'rgba(0, 0, 0, 0.1)' }}
        markStyle={{ color: 'red' }}
      >
        <div style={{ height: 500 }}>Content</div>
      </WaterMark>,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders without content when content prop is not provided', () => {
    const { container } = render(
      <WaterMark>
        <div style={{ height: 500 }}>Content</div>
      </WaterMark>,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders without watermark when both image and content props are not provided', () => {
    const { container } = render(
      <WaterMark>
        <div style={{ height: 500 }}>Content</div>
      </WaterMark>,
    );
    expect(container).toMatchSnapshot();
  });
});
