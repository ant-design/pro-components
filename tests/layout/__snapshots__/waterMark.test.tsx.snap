// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WaterMark > renders watermark with custom styles 1`] = `
<div>
  <div
    class="ant-pro-layout-watermark-wrapper"
    style="position: relative; background-color: rgba(0, 0, 0, 0.1);"
  >
    <div
      style="height: 500px;"
    >
      Content
    </div>
    <div
      class="ant-pro-layout-watermark"
      style="z-index: 9; position: absolute; left: 0px; top: 0px; width: 100%; height: 100%; background-size: 332px; pointer-events: none; background-repeat: repeat; color: red;"
    />
  </div>
</div>
`;

exports[`WaterMark > renders watermark with multiline text content 1`] = `
<div>
  <div
    class="ant-pro-layout-watermark-wrapper"
    style="position: relative;"
  >
    <div
      style="height: 500px;"
    >
      Content
    </div>
    <div
      class="ant-pro-layout-watermark"
      style="z-index: 9; position: absolute; left: 0px; top: 0px; width: 100%; height: 100%; background-size: 332px; pointer-events: none; background-repeat: repeat;"
    />
  </div>
</div>
`;

exports[`WaterMark > renders watermark with rotation 1`] = `
<div>
  <div
    class="ant-pro-layout-watermark-wrapper"
    style="position: relative;"
  >
    <div
      style="height: 500px;"
    >
      Content
    </div>
    <div
      class="ant-pro-layout-watermark"
      style="z-index: 9; position: absolute; left: 0px; top: 0px; width: 100%; height: 100%; background-size: 332px; pointer-events: none; background-repeat: repeat;"
    />
  </div>
</div>
`;

exports[`WaterMark > renders without content when content prop is not provided 1`] = `
<div>
  <div
    class="ant-pro-layout-watermark-wrapper"
    style="position: relative;"
  >
    <div
      style="height: 500px;"
    >
      Content
    </div>
    <div
      class="ant-pro-layout-watermark"
      style="z-index: 9; position: absolute; left: 0px; top: 0px; width: 100%; height: 100%; background-size: 332px; pointer-events: none; background-repeat: repeat;"
    />
  </div>
</div>
`;

exports[`WaterMark > renders without watermark when both image and content props are not provided 1`] = `
<div>
  <div
    class="ant-pro-layout-watermark-wrapper"
    style="position: relative;"
  >
    <div
      style="height: 500px;"
    >
      Content
    </div>
    <div
      class="ant-pro-layout-watermark"
      style="z-index: 9; position: absolute; left: 0px; top: 0px; width: 100%; height: 100%; background-size: 332px; pointer-events: none; background-repeat: repeat;"
    />
  </div>
</div>
`;

exports[`WaterMark > test image watermark 1`] = `
<div>
  <div
    class="ant-pro-layout-watermark-wrapper"
    style="position: relative;"
  >
    <div
      style="height: 500px;"
    >
      123
    </div>
    <div
      class="ant-pro-layout-watermark"
      style="z-index: 9; position: absolute; left: 0px; top: 0px; width: 100%; height: 100%; background-size: 332px; pointer-events: none; background-repeat: repeat;"
    />
  </div>
</div>
`;

exports[`WaterMark > test text watermark 1`] = `
<DocumentFragment>
  <div
    class="ant-pro-layout-watermark-wrapper"
    style="position: relative;"
  >
    <div
      style="height: 500px;"
    />
    <div
      class="ant-pro-layout-watermark"
      style="z-index: 9; position: absolute; left: 0px; top: 0px; width: 100%; height: 100%; background-size: 332px; pointer-events: none; background-repeat: repeat;"
    />
  </div>
</DocumentFragment>
`;
